### PR TITLE
add validation for ClusterAffinities

### PR DESCRIPTION
**Pull request by XiShanYongYe-Chang**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Part of #3153 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
`karmada-webhook`: add validation for policy.spec.placement.orderedClusterAffinities
```

### CODE REVIEW

File: pkg/util/validation/validation.go [hunk 1]
```diff
@@ -16,6 +16,67 @@ import (
...
+}
+
+// ValidateOrderedClusterAffinities validates a webhook before creation or update.
+func ValidateOrderedClusterAffinities(affinities []policyv1alpha1.OrderedClusterAffinityTerm, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	orders := make(map[int32]bool)
+	for index, term := range affinities {
+		if term.Order < clusterGroupMinOrder || term.Order > clusterGroupMaxOrder {
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(index).Child("order"), term.Order, "order need to be in the range 1-100"))
```
Comment (chaunceyjiang) [inline pkg/util/validation/validation.go:66]: Why not use `+kubebuilder:validation:Maximum=100` and `+kubebuilder:validation: Minimum=1`  ?

File: pkg/util/validation/validation.go [hunk 2]
```diff
@@ -40,18 +101,19 @@ func ValidatePolicyFieldSelector(fieldSelector *policyv1alpha1.FieldSelector) er
 }
 
 // ValidateSpreadConstraint tests if the constraints is valid.
-func ValidateSpreadConstraint(spreadConstraints []policyv1alpha1.SpreadConstraint) error {
-	spreadByFields := sets.NewString()
+func ValidateSpreadConstraint(spreadConstraints []policyv1alpha1.SpreadConstraint, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
 
-	for _, constraint := range spreadConstraints {
+	spreadByFields := sets.NewString()
```
Comment (chaunceyjiang) [inline pkg/util/validation/validation.go:107]:
```suggestion
	spreadByFields := sets.New[string]()
```

Comment (XiShanYongYe-Chang) [reply pkg/util/validation/validation.go]: updated

File: pkg/util/validation/validation.go (hunk 1 above)
Comment (XiShanYongYe-Chang) [reply pkg/util/validation/validation.go]: I think we can do what you say.

**Comment by XiShanYongYe-Chang** [comment]
/cc @chaunceyjiang 
Can you help review it again?

**Review by chaunceyjiang — COMMENTED** [review]
/lgtm

File: pkg/util/validation/validation.go [hunk 3]
```diff
@@ -16,6 +16,58 @@ import (
...
+}
+
+// ValidateOrderedClusterAffinities validates a webhook before creation or update.
+func ValidateOrderedClusterAffinities(affinities []policyv1alpha1.ClusterAffinityTerm, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	groupNames := make(map[string]bool)
+	for index, term := range affinities {
+		if groupNames[term.AffinityName] {
+			allErrs = append(allErrs, field.Invalid(fldPath, affinities, "different cluster group can not have the same affinityName"))
```
Comment (RainbowMango) [inline pkg/util/validation/validation.go:61]:
```suggestion
			allErrs = append(allErrs, field.Invalid(fldPath, affinities, "each affinity term in a policy must have a unique name"))
```

File: pkg/util/validation/validation.go [hunk 4]
```diff
@@ -16,6 +16,58 @@ import (
...
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("fieldSelector"), affinity.FieldSelector, err.Error()))
+	}
+	return allErrs
+}
+
+// ValidateOrderedClusterAffinities validates a webhook before creation or update.
+func ValidateOrderedClusterAffinities(affinities []policyv1alpha1.ClusterAffinityTerm, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	groupNames := make(map[string]bool)
```
Comment (RainbowMango) [inline pkg/util/validation/validation.go:58]:
```suggestion
	affinityNames := make(map[string]bool)
```

File: pkg/util/validation/validation.go [hunk 5]
```diff
@@ -16,6 +16,58 @@ import (
...
+	for index, term := range affinities {
+		if groupNames[term.AffinityName] {
```
Comment (RainbowMango) [inline pkg/util/validation/validation.go:60]:
```suggestion
		if _, exist := groupNames[term.AffinityName]; exist {
```

File: pkg/util/validation/validation.go [hunk 6]
```diff
@@ -16,6 +16,58 @@ import (
 // LabelValueMaxLength is a label's max length
 const LabelValueMaxLength int = 63
 
+// ValidatePropagationSpec validates a webhook before creation or update.
```
[the remark this answers is not in the document]
Comment (XiShanYongYe-Chang) [reply pkg/util/validation/validation.go]: Update it to `PropagationSpec`

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**